File: Challenge-005-Queue/C++/Queue.cpp
#include "Queue.h"

void Queue::enqueue (func_ptr data)
{
    if(!isFull())
    {
        front+=1;
        items[front] = data;
    }
}

void Queue::dequeue (void)
{
    if(!isEmpty())
    {
        items[FRONT]();
        for(int i=FRONT;i<front;i++)
        {
            items[i] = items[i+1];
        }
        front-=1;
    }
}

void Queue::peak    (void)
{
    if(!isEmpty())
    {
        items[FRONT]();
    }
}

int Queue::isFull  (void)
{
    return (MAX_QUEUE_SIZE==front)?1:0;
}

int Queue::isEmpty (void)
{
    return (EMPTY==front)?1:0;
}

File: Challenge-005-Queue/java/Queue.java
public class Queue{
    private static final int MAX_SIZE = 100;
    private int[] items = new int[MAX_SIZE];
    private int front = 0;
    private int rear = -1;

    public boolean isEmpty() {
        return rear < front;
    }

    public boolean isFull() {
        return rear - front + 1 == MAX_SIZE;
    }

    public void enqueue(int value) {
        if (isFull()) {
            System.out.println("Queue overflow");
            return;
        }
        items[++rear] = value;
    }

    public int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue underflow");
            return -1;
        }
        return items[front++];
    }

    public int peek() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
            return -1;
        }
        return items[front];
    }
}
